### ISSUE TITLE
Custom component setup.py in kubeflow pipeline

### ISSUE THREAD

**Issue by DvirDukhan** [tensorflow/tfx#2549]
Hi
I'm trying to build a custom example generator that reads from Redis and deploy it in kubeflow.
https://github.com/DvirDukhan/redis_example_gen
I'm using kubeflow pipelines over GKE cluster and trying to modify the taxi demo template to use my example generator.
I added the component logic folder and proto-buf definitions to the pipeline folder in the template and added its `setup.py` file to the pipeline root.
```shell
.
├── configs.py
├── __init__.py
├── pipeline.py
├── redis_component
│   ├── component.py
│   ├── executor.py
│   ├── __init__.py
│   └── version.py
├── redis_proto
│   ├── __init__.py
│   ├── redis_config.proto
│   └── redis_hash_query.proto
└── setup.py
```

modified the `pipeline.py` file to use my component (its inputs are connection configuration and a query)
```python
  example_gen = RedisExampleGen(conn_config=_redis_config, query=_query)
```

I modified the `configs.py` file to have the beam pipeline args point the setup file and modified the `kubeflow_dag_runner.py` file to use this config. The location is pointed since the generated dockerfile sets the `WORKDIR` to `/pipeline`
```python
READIS_BEAM_PIPELINE_ARGS = [
    '--setup_file=/pipeline/setup.py'
]
```
```python
      pipeline.create_pipeline(
          pipeline_name=configs.PIPELINE_NAME,
          pipeline_root=PIPELINE_ROOT,
          data_path=DATA_PATH,
          # TODO(step 7): (Optional) Uncomment below to use BigQueryExampleGen.
          # query=configs.BIG_QUERY_QUERY,
          preprocessing_fn=configs.PREPROCESSING_FN,
          run_fn=configs.RUN_FN,
          train_args=trainer_pb2.TrainArgs(num_steps=configs.TRAIN_NUM_STEPS),
          eval_args=trainer_pb2.EvalArgs(num_steps=configs.EVAL_NUM_STEPS),
          eval_accuracy_threshold=configs.EVAL_ACCURACY_THRESHOLD,
          serving_model_dir=SERVING_MODEL_DIR,
          beam_pipeline_args=configs
          .REDIS_BEAM_PIPELINE_ARGS,
          # TODO(step 7): (Optional) Uncomment below to use provide GCP related
          #               config for BigQuery with Beam DirectRunner.
          # beam_pipeline_args=configs
          # .BIG_QUERY_WITH_DIRECT_RUNNER_BEAM_PIPELINE_ARGS,
          # TODO(step 8): (Optional) Uncomment below to use Dataflow.
          # beam_pipeline_args=configs.DATAFLOW_BEAM_PIPELINE_ARGS,
          # TODO(step 9): (Optional) Uncomment below to use Cloud AI Platform.
          # ai_platform_training_args=configs.GCP_AI_PLATFORM_TRAINING_ARGS,
          # TODO(step 9): (Optional) Uncomment below to use Cloud AI Platform.
          # ai_platform_serving_args=configs.GCP_AI_PLATFORM_SERVING_ARGS,
      ))

```

Dockerfile:
```
FROM tensorflow/tfx:0.23.0
WORKDIR /pipeline
COPY ./ ./
ENV PYTHONPATH="/pipeline:${PYTHONPATH}"
```

I managed to deploy the pipeline, however, I cannot execute an experiment as the component is not installed.

```bash
2020-09-22 21:44:29.991281: W tensorflow/stream_executor/platform/default/dso_loader.cc:59] Could not load dynamic library 'libcudart.so.10.1'; dlerror: libcudart.so.10.1: cannot open shared object file: No such file or directory; LD_LIBRARY_PATH: /usr/local/lib
2020-09-22 21:44:29.991331: I tensorflow/stream_executor/cuda/cudart_stub.cc:29] Ignore above cudart dlerror if you do not have a GPU set up on your machine.
Traceback (most recent call last):
  File "/tfx-src/tfx/orchestration/kubeflow/container_entrypoint.py", line 360, in <module>
    main()
  File "/tfx-src/tfx/orchestration/kubeflow/container_entrypoint.py", line 315, in main
    component = json_utils.loads(args.serialized_component)
  File "/tfx-src/tfx/utils/json_utils.py", line 193, in loads
    return json.loads(s, cls=_DefaultDecoder)
  File "/usr/lib/python3.7/json/__init__.py", line 361, in loads
    return cls(**kw).decode(s)
  File "/usr/lib/python3.7/json/decoder.py", line 337, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/usr/lib/python3.7/json/decoder.py", line 353, in raw_decode
    obj, end = self.scan_once(s, idx)
  File "/tfx-src/tfx/utils/json_utils.py", line 174, in _dict_to_object
    return _extract_class(dict_data)
  File "/tfx-src/tfx/utils/json_utils.py", line 164, in _extract_class
    return getattr(importlib.import_module(module_name), class_name)
  File "/usr/lib/python3.7/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1006, in _gcd_import
  File "<frozen importlib._bootstrap>", line 983, in _find_and_load
  File "<frozen importlib._bootstrap>", line 953, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "<frozen importlib._bootstrap>", line 1006, in _gcd_import
  File "<frozen importlib._bootstrap>", line 983, in _find_and_load
  File "<frozen importlib._bootstrap>", line 965, in _find_and_load_unlocked
ModuleNotFoundError: No module named 'redis_component'
```

I tried to change the beam setup file parameter to point to different (reasonable) locations as `./setup.py` etc.. but without success. 
My question is simple: Are there any additional configurations or actions during the pipeline build that are required for the pipeline run time environment to install the custom component?

**Comment by lukemarsden** [comment]
I've seen this approach used to install dependencies in KFP, it's not pretty, but it might get you unblocked: https://github.com/pachyderm/pachyderm/blob/master/examples/kubeflow/mnist/pipeline.py#L17-L20

Maybe some of the maintainers can advise on a better way to do it, and/or why `--setup_file` isn't working. (My only guesses there are: maybe it's `--setup-file`? Or maybe `READIS_BEAM_PIPELINE_ARGS` has a typo `READIS`? But those are total guesses, I don't know the context).

**Comment by DvirDukhan** [comment]
Thanks, @lukemarsden for the answer
1. My apologies for the typo on `REDIS_BEAM_PIPELINE_ARGS` on the issue description. The issue still remains after I verified on `config.py`
```python
REDIS_BEAM_PIPELINE_ARGS = [
    '--setup_file=/pipeline/setup.py'
]
```
and on `kubeflow_dag_runner.py` 
```python
beam_pipeline_args=configs
.REDIS_BEAM_PIPELINE_ARGS,
```
2. `--setup_file` is the right flag, as documented [here](https://beam.apache.org/documentation/sdks/python-pipeline-dependencies/)
3. I tried to implement the approach of installing the package directly from the pipeline code, without success.
Thank you for your reply. I will wait for answers from the maintainers.
